internal/pkg/image/packer/gocryptfs.go

```diff
@@ -0,0 +1,194 @@
+package packer
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/apptainer/apptainer/internal/pkg/image/driver"
+	"github.com/apptainer/apptainer/internal/pkg/util/bin"
+	"github.com/apptainer/apptainer/pkg/sylog"
+)
+
+type Gocryptfs struct {
+	*Squashfs
+	gocryptfsPath  string
+	fusermountPath string
+	Pass           string
+}
+
+type cryptInfo struct {
+	cipherDir, plainDir, pass, confPath, parentDir string
+	pid                                            int
+}
+
+var defaultCleanFnGenerator = func(target string, pid int) func(bool) {
+	return func(kill bool) {
+		// if using syscall.Unmount, it'll show permission denied error
+		cmd := exec.Command("fusermount", "-u", target)
```

Comment:
This is a big red flag to me.  fusermount is a setuid-root program.  We should not have to be using that.  It should be happening at a point where we can use a root-mapped unprivileged user namespace to mount and unmount.

Reply:
here, it looks like we are in the host namespace, and gocryptfs itself will also call fusermount to mount fuse fs. so in either way, fusermount will be called.

Reply:
That's probably true, but using fusermount to do the fuse mount (in addition to the fact that it can't be done in a root-restricted environment such as inside an apptainer container) means that cleanup can be a mess, for example if it is killed with `kill -9` (or maybe even a `^C`, you might want to try that) then there will be a leftover mount that's connected to nothing which will get `Transport endpoint is not connected` errors.
 
I think it should be possible to do the original mount in a new root-mapped unprivileged user+mount namespace.  We probably could even make it a pid namespace as well which will make sure that the FUSE process exits when the namespace is shut down.   The [`fakeroot.UnshareRootMapped()`](https://github.com/apptainer/apptainer/blob/release-1.1/internal/pkg/fakeroot/fakefake.go#L29) function may be close to what you need, and I am about to modify it to add a mount namespace so it can bind mount the home directory.  You would need to make sure the same namespace stays in place for the duration of the command, not just execute a subcommand in it, so the mountpoint stays available to the rest of the code.  The way that function is designed to be used is to pass `os.Args` to it like [here](https://github.com/apptainer/apptainer/blob/main/internal/pkg/runtime/launch/launcher_linux.go#L123) so it re-executes the whole program from scratch.  The next time through it detects that things are already set up and then skips the re-exec.

Reply:
yeah, I tested pressing `^C` when building a sif image, and it'll leave an orphan gocryptfs process.  Yeah, we should put the original mount inside a new namespace.

Reply:
#1230 now has a modified `UnshareRootMapped()` function that always adds a mount namespace in addition to a root namespace.  So I think with that when in fakeroot mode (either with /etc/subuid mapping or not) you should be able to start gocryptfs without it invoking fusermount.  Since we're talking about building containers it is always in fakeroot mode, so that may already take care of the trickiest part.

Reply:
thanks! I've refactored the code and use namespace for the build step.
But one test in the e2e test fails
```
2023-03-27T07:05:50.8786252Z === CONT  TestE2E/PAR/RUN/add_package_with_fakeroot_and_tmpfs
2023-03-27T07:05:50.8787895Z     apptainercmd.go:694: Running command "/usr/local/bin/apptainer exec --fakeroot --userns --writable-tmpfs --ignore-subuid --ignore-fakeroot-command /tmp/stest.3565202831/test-debian.sif sh -c apt-get update && apt-get install -y openssh-client"
2023-03-27T07:05:52.3119905Z     apptainercmd.go:694: 
2023-03-27T07:05:52.3121653Z         "/usr/local/bin/apptainer exec --fakeroot --userns --writable-tmpfs --ignore-subuid --ignore-fakeroot-command /tmp/stest.3565202831/test-debian.sif sh -c apt-get update && apt-get install -y openssh-client" output:
2023-03-27T07:05:52.3122671Z         INFO:    User not listed in /etc/subuid, trying root-mapped namespace
2023-03-27T07:05:52.3303443Z         INFO:    fakeroot command not found, using only root-mapped namespace
2023-03-27T07:05:52.3304354Z         E: setgroups 65534 failed - setgroups (1: Operation not permitted)
2023-03-27T07:05:52.3305102Z         E: setegid 65534 failed - setegid (22: Invalid argument)
2023-03-27T07:05:52.3305746Z         E: seteuid 100 failed - seteuid (22: Invalid argument)
2023-03-27T07:05:52.3306356Z         E: setgroups 0 failed - setgroups (1: Operation not permitted)
2023-03-27T07:05:52.3351223Z         W: chown to _apt:root of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (22: Invalid argument)
2023-03-27T07:05:52.3352103Z         W: chown to _apt:root of directory /var/lib/apt/lists/auxfiles failed - SetupAPTPartialDirectory (22: Invalid argument)
2023-03-27T07:05:52.3352730Z         E: setgroups 65534 failed - setgroups (1: Operation not permitted)
2023-03-27T07:05:52.3353226Z         E: setegid 65534 failed - setegid (22: Invalid argument)
2023-03-27T07:05:52.3353702Z         E: seteuid 100 failed - seteuid (22: Invalid argument)
2023-03-27T07:05:52.3354224Z         E: setgroups 0 failed - setgroups (1: Operation not permitted)
2023-03-27T07:05:52.3354946Z         E: Method gave invalid 400 URI Failure message: Failed to setgroups - setgroups (1: Operation not permitted)
2023-03-27T07:05:52.3355398Z         E: Method http has died unexpectedly!
2023-03-27T07:05:52.3355863Z         E: Sub-process http returned an error code (112)
2023-03-27T07:05:52.3356580Z         FATAL:   --fakeroot requires either being in /etc/subuid, unprivileged user namespaces, or the fakeroot command
2023-03-27T07:05:52.3357206Z         Reading package lists...
2023-03-27T07:05:52.3357432Z         
2023-03-27T07:05:52.3357881Z     apptainercmd.go:694: got 255 as exit code and was expecting 100: exit status 255
```
not sure if it is related with the gocryptfs changes.

Reply:
That test is actually expected to fail with similar messages.  The difference is that it used to fail with a 100 code, and now it is failing with a FATAL message
```
FATAL:   --fakeroot requires either being in /etc/subuid, unprivileged user namespaces, or the fakeroot command
```
which returns a 255 error code.  It's very odd to get a fatal error at that point, after running the command.  I can reproduce interactively with this branch and
```
apptainer exec --fakeroot --userns --writable-tmpfs --ignore-subuid --ignore-fakeroot-command docker://ubuntu:20.04 apt-get update
```

Reply:
> That test is actually expected to fail with similar messages. The difference is that it used to fail with a 100 code, and now it is failing with a FATAL message
> 
> ```
> FATAL:   --fakeroot requires either being in /etc/subuid, unprivileged user namespaces, or the fakeroot command
> ```
> 
> which returns a 255 error code. It's very odd to get a fatal error at that point, after running the command. I can reproduce interactively with this branch and
> 
> ```
> apptainer exec --fakeroot --userns --writable-tmpfs --ignore-subuid --ignore-fakeroot-command docker://ubuntu:20.04 apt-get update
> ```

thanks for the help! it is introduced by my recent commit
https://github.com/apptainer/apptainer/pull/1185/commits/d1a9600aa333cc64012c933c62f7b4c372d93760
where one error is returned rather than just printed.  Never mind, I've reverted the change.

Reply:
> Actually, I was hoping you wouldn't need to add an additional call to `UnshareRootMapped()` at all. After the inclusion of #1230 (which is now merged, rebase on the main branch please) I think that you should be able to do a fuse mount anytime that fakeroot is enabled, either via /etc/subuid or via `UnshareRootMapped()`. So if you start gocryptfs after the point that fakeroot is processed, either way, then I think it should just work without invoking fusermount.

Getting a bit of confused here, if we are creating sif based on oci image or existing plain image, it won't run in `--fakeroot` mode? only if we are creating sif using definition file, then it'll run in `--fakeroot` mode automatically.  So we need to initialize gocryptfs fuse mount in two different places?  In terms of suid mode, we currently need to add two options `"--ignore-subuid", "--userns"` because imagedriver is not initialized in suid mode.

Reply:
I've refactored the code and also make fakeroot mode always enabled when building images.
https://github.com/apptainer/apptainer/blob/2c05f5dd25f2efe9b3051b923b6d07b38a4982ea/cmd/internal/cli/build.go#L351

sorry, my fault, we just need use `--userns`,  no need to use `--ignore-subuid `.